// @ngrx/store/esm5/src/tokens.js
import { InjectionToken } from '@angular/core';
export var _INITIAL_STATE = new InjectionToken('@ngrx/store Internal Initial State');
export var INITIAL_STATE = new InjectionToken('@ngrx/store Initial State');
export var REDUCER_FACTORY = new InjectionToken('@ngrx/store Reducer Factory');
export var _REDUCER_FACTORY = new InjectionToken('@ngrx/store Reducer Factory Provider');
export var INITIAL_REDUCERS = new InjectionToken('@ngrx/store Initial Reducers');
export var _INITIAL_REDUCERS = new InjectionToken('@ngrx/store Internal Initial Reducers');
export var META_REDUCERS = new InjectionToken('@ngrx/store Meta Reducers');
export var STORE_FEATURES = new InjectionToken('@ngrx/store Store Features');
export var _STORE_REDUCERS = new InjectionToken('@ngrx/store Internal Store Reducers');
export var _FEATURE_REDUCERS = new InjectionToken('@ngrx/store Internal Feature Reducers');
export var _FEATURE_REDUCERS_TOKEN = new InjectionToken('@ngrx/store Internal Feature Reducers Token');
export var FEATURE_REDUCERS = new InjectionToken('@ngrx/store Feature Reducers');
